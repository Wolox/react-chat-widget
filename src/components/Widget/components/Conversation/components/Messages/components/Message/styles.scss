@import 'variables.scss';
@import 'common.scss';

$fontName : productsans;
@font-face {
  font-family: $fontName;
  src: url('/assets/Product-Sans-Regular.ttf');
}
.rcw-message {
  margin: 10px;
  display: flex;
  word-wrap: break-word;
  word-break: break-word;
   font-family: $fontName;
}

.rcw-client {
  @include message-bubble($g-red);
  background: -webkit-linear-gradient(-29deg, #00d8d4, #00ef70) !important;
  margin-left: auto;
}

.rcw-response {
  @include message-bubble($grey-2);
}

/* For markdown elements created with default styles */
.rcw-message-text {
  font-family: GoogleSans, Helvetica, Arial, sans-serif;
  p {
    margin: 0;
  }
   
}

.rcw-client .rcw-message-text {
  color: white;
  font-family: GoogleSans, Helvetica, Arial, sans-serif;

}

.rcw-avatar {
  width: 40px;
  height: 40px;
  border-radius: 100%;
  margin-right: 10px;
}
