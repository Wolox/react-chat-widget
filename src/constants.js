import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';

export const MESSAGE_SENDER = {
  CLIENT: 'client',
  RESPONSE: 'response'
};

export const MESSAGES_TYPES = {
  TEXT: 'text',
  SNIPPET: {
    LINK: 'snippet'
  },
  CUSTOM_COMPONENT: 'component'
};

export const PROP_TYPES = {
  MESSAGE: ImmutablePropTypes.contains({
    type: PropTypes.oneOf([
      MESSAGES_TYPES.TEXT,
      MESSAGES_TYPES.SNIPPET.LINK
    ]),
    text: PropTypes.string,
    sender: PropTypes.oneOf([
      MESSAGE_SENDER.CLIENT,
      MESSAGE_SENDER.RESPONSE
    ])
  }),

  SNIPPET: ImmutablePropTypes.contains({
    type: PropTypes.oneOf([
      MESSAGES_TYPES.TEXT,
      MESSAGES_TYPES.SNIPPET.LINK
    ]),
    title: PropTypes.string,
    link: PropTypes.string,
    sender: PropTypes.oneOf([
      MESSAGE_SENDER.CLIENT,
      MESSAGE_SENDER.RESPONSE
    ])
  })
};

export const MESSAGE_BOX_SCROLL_DURATION = 400;
