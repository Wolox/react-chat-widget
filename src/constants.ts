export enum MessageOrigin {
  client = 'client',
  response = 'response'
};

export const MESSAGES_TYPES = {
  TEXT: 'text',
  SNIPPET: {
    LINK: 'snippet'
  },
  CUSTOM_COMPONENT: 'component'
};

export const MESSAGE_BOX_SCROLL_DURATION = 400;
