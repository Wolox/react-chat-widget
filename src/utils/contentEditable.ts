export const getCaretIndex = (el: HTMLElement) => {
  let position = 0;
  const selection = window.getSelection()!;
  if (selection.rangeCount !== 0) {
    const range = window.getSelection()!.getRangeAt(0);
    const preCaretRange = range.cloneRange();
    preCaretRange.selectNodeContents(el);
    preCaretRange.setEnd(range.endContainer, range.endOffset);
    position = preCaretRange.toString().length;
  }
  return position;
}

export const getSelection = (el: HTMLElement) => {
  const range = window.getSelection()!.getRangeAt(0);
  const preSelectionRange = range.cloneRange();
  preSelectionRange.selectNodeContents(el);
  preSelectionRange.setEnd(range.startContainer, range.startOffset);

  const start = preSelectionRange.toString().length;
  return {
    start: start,
    end: start + range.toString().length
  }
}

export const isFirefox = () => navigator.userAgent.search("Firefox") > 0;

export const updateCaret = (el: HTMLElement, caret: number, offset: number) => {
  const range = document.createRange();
  const selection = window.getSelection()!;
  range.setStart(el.childNodes[0], caret + offset);
  range.collapse(true);
  selection.removeAllRanges();
  selection.addRange(range);
  el.focus();
}

export const insertNodeAtCaret = (el: HTMLElement) => {
  const position = getCaretIndex(el)
  let characterToEnter = '\n\n';
  let prevChar, char = '';
  if (position > 0) {
    prevChar = el.innerHTML.charAt(position - 1);
    char = el.innerHTML.charAt(position);
    const newLines = el.innerHTML.match(/\n/g);
    if (
      prevChar === char ||
      (prevChar === '\n' && char === '') ||
      (isFirefox() && (newLines?.length || 0) > 0)
    ) {
      characterToEnter = '\n';
    }
  }
  const selection = window.getSelection()!;
  const node = document.createTextNode(characterToEnter);
  const range = selection.getRangeAt(0);
  range.collapse(false);
  range.insertNode(node);
  const cloneRange = range.cloneRange();
  cloneRange.selectNodeContents(node);
  cloneRange.collapse(false);
  selection.removeAllRanges();
  selection.addRange(cloneRange);
  el.innerHTML = el.innerHTML.replace(/<br>/g, '');
  updateCaret(el, position, 1);
}
