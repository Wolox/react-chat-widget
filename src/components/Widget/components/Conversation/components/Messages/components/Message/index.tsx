import format from 'date-fns/format';
import MarkdownIt from 'markdown-it';
import React from 'react';
import { MessageTypes } from '../../../../../../../../store/types';
import './styles.scss';



type Props = {
  message: MessageTypes;
  showTimeStamp: boolean;
}

function Message({ message, showTimeStamp }: Props) {
  const sanitizedHTML = new MarkdownIt("default", { linkify: false })
    .render(message.text);

  return (
    <div className={`rcw-${message.sender}`}>
      <div className="rcw-message-text" dangerouslySetInnerHTML={{ __html: sanitizedHTML.replace(/\n$/, '') }} />
      {showTimeStamp && <span className="rcw-timestamp">{format(message.timestamp, 'hh:mm')}</span>}
    </div>
  );
}

export default Message;
