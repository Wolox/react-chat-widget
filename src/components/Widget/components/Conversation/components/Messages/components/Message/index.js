import React, {PureComponent} from 'react';
import markdownIt from 'markdown-it';
import markdownItSup from 'markdown-it-sup';
import markdownItSanitizer from 'markdown-it-sanitizer';
import markdownItLinkAttributes from 'markdown-it-link-attributes';

import {PROP_TYPES} from './src/constants.js';

import './styles.scss';

class Message extends PureComponent {
  render() {
    const sanitizedHTML = markdownIt()
      .use(markdownItSup)
      .use(markdownItSanitizer)
      .use(markdownItLinkAttributes, {
        attrs: {target: '_blank', rel: 'noopener'},
      })
      .render(this.props.message.get('text'));

    return (
      <div className={`rcw-${this.props.message.get('sender')}`}>
        <div
          className="rcw-message-text"
          dangerouslySetInnerHTML={{__html: sanitizedHTML}}
        />
      </div>
    );
  }
}

Message.propTypes = {
  message: PROP_TYPES.MESSAGE,
};

export default Message;
