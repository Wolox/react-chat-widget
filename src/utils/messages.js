import { Map } from 'immutable';
import { MESSAGES_TYPES, MESSAGE_SENDER, MESSAGE_BOX_SCROLL_DURATION } from '@constants';

import Message from '@messagesComponents/Message';
import Snippet from '@messagesComponents/Snippet';
import QuickButton from '@quickButtonsComponents/QuickButton';

export function createNewMessage(text, sender) {
  return Map({
    type: MESSAGES_TYPES.TEXT,
    component: Message,
    text,
    sender,
    showAvatar: sender === MESSAGE_SENDER.RESPONSE
  });
}

export function createLinkSnippet(link) {
  return Map({
    type: MESSAGES_TYPES.SNIPPET.LINK,
    component: Snippet,
    title: link.title,
    link: link.link,
    target: link.target || '_blank',
    sender: MESSAGE_SENDER.RESPONSE,
    showAvatar: true
  });
}

export function createComponentMessage(component, props, showAvatar) {
  return Map({
    type: MESSAGES_TYPES.CUSTOM_COMPONENT,
    component,
    props,
    sender: MESSAGE_SENDER.RESPONSE,
    showAvatar
  });
}

/**
 * Easing Functions
 * @param {*} t timestamp
 * @param {*} b begining
 * @param {*} c change
 * @param {*} d duration
 */
function sinEaseOut(t, b, c, d) {
  return c * ((t = t / d - 1) * t * t + 1) + b;
}

/**
 * 
 * @param {*} target scroll target
 * @param {*} scrollStart
 * @param {*} scroll scroll distance
 */
function scrollWithSlowMotion(target, scrollStart, scroll) {
  const raf = window.webkitRequestAnimationFrame || window.requestAnimationFrame
  let start = null
  const step = (timestamp) => {
    if (!start) {
      start = timestamp
    }
    let stepScroll = sinEaseOut(timestamp - start, 0, scroll, MESSAGE_BOX_SCROLL_DURATION)
    let total = scrollStart + stepScroll
    target.scrollTop = total;
    if (total < scrollStart + scroll) {
      raf(step)
    }
  }
  raf(step)
}

export function scrollToBottom(messagesDiv) {
  if (!messagesDiv) return;
  const screenHeight = messagesDiv.clientHeight;
  const scrollTop = messagesDiv.scrollTop;

  const scrollOffset = messagesDiv.scrollHeight - (scrollTop + screenHeight)

  scrollOffset && scrollWithSlowMotion(messagesDiv, scrollTop, scrollOffset);
}


export function createQuickButton(button) {
  return Map({
    component: QuickButton,
    label: button.label,
    value: button.value
  });
}
