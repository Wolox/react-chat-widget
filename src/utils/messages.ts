import { ElementType } from 'react';

import { MessageTypes as MessageI, Link, CustomCompMessage, LinkParams } from '../store/types';

import Message from '../components/Widget/components/Conversation/components/Messages/components/Message';
import Snippet from '../components/Widget/components/Conversation/components/Messages/components/Snippet';
import QuickButton from '../components/Widget/components/Conversation/components/QuickButtons/components/QuickButton';

import { MESSAGES_TYPES, MESSAGE_SENDER, MESSAGE_BOX_SCROLL_DURATION } from '../constants';

export function createNewMessage(
  text: string,
  sender: string,
  id?: string,
): MessageI {
  return {
    type: MESSAGES_TYPES.TEXT,
    component: Message,
    text,
    sender,
    timestamp: new Date(),
    showAvatar: true,
    customId: id,
    unread: sender === MESSAGE_SENDER.RESPONSE
  };
}

export function createLinkSnippet(link: LinkParams, id?: string): Link {
  return {
    type: MESSAGES_TYPES.SNIPPET.LINK,
    component: Snippet,
    title: link.title,
    link: link.link,
    target: link.target || '_blank',
    sender: MESSAGE_SENDER.RESPONSE,
    timestamp: new Date(),
    showAvatar: true,
    customId: id,
    unread: true
  };
}

export function createComponentMessage(component: ElementType, props: any, showAvatar: boolean, id?: string): CustomCompMessage {
  return {
    type: MESSAGES_TYPES.CUSTOM_COMPONENT,
    component,
    props,
    sender: MESSAGE_SENDER.RESPONSE,
    timestamp: new Date(),
    showAvatar,
    customId: id,
    unread: true
  };
}

export function createQuickButton(button: { label: string, value: string | number }) {
  return {
    component: QuickButton,
    label: button.label,
    value: button.value
  };
}

// TODO: Clean functions and window use for SSR

function sinEaseOut(timestamp: any, begining: any, change: any, duration: any) {
  return change * ((timestamp = timestamp / duration - 1) * timestamp * timestamp + 1) + begining;
}

/**
 * 
 * @param {*} target scroll target
 * @param {*} scrollStart
 * @param {*} scroll scroll distance
 */
function scrollWithSlowMotion(target: any, scrollStart: any, scroll: number) {
  const raf = window?.requestAnimationFrame;
  let start = 0;
  const step = (timestamp: number) => {
    if (!start) {
      start = timestamp;
    }
    let stepScroll = sinEaseOut(timestamp - start, 0, scroll, MESSAGE_BOX_SCROLL_DURATION);
    let total = scrollStart + stepScroll;
    target.scrollTop = total;
    if (total < scrollStart + scroll) {
      raf(step);
    }
  }
  raf(step);
}

export function scrollToBottom(messagesDiv: HTMLDivElement | null) {
  if (!messagesDiv) return;
  const screenHeight = messagesDiv.clientHeight;
  const scrollTop = messagesDiv.scrollTop;
  const scrollOffset = messagesDiv.scrollHeight - (scrollTop + screenHeight);
  if (scrollOffset) scrollWithSlowMotion(messagesDiv, scrollTop, scrollOffset);
}
