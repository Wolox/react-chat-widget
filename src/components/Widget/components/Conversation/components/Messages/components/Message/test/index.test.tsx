import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { MessageOrigin } from '../../../../../../../../../constants';
import { createNewMessage } from '../../../../../../../../../utils/messages';
import Message from '../index';

configure({ adapter: new Adapter() });

describe('<Message />', () => {
  /* eslint-disable no-underscore-dangle */
  const createMessageComponent = message => shallow(<Message message={message} showTimeStamp timestampFormat="hh:mm" />);

  it('should render a <strong> element', () => {
    const message = createNewMessage('New message with **Markdown**!', MessageOrigin.client);
    const messageComponent = createMessageComponent(message);
    expect(messageComponent.find('.rcw-message-text').getElement().props.dangerouslySetInnerHTML.__html).toMatchSnapshot();
  });

  it('should reder a <em> element', () => {
    const message = createNewMessage('New message with *Markdown*!', MessageOrigin.client);
    const messageComponent = createMessageComponent(message);
    expect(messageComponent.find('.rcw-message-text').getElement().props.dangerouslySetInnerHTML.__html).toMatchSnapshot();
  });
  /* eslint-enable */
});
