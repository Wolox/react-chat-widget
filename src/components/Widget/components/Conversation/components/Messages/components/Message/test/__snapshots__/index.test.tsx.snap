// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message /> should reder a <em> element 1`] = `"<p>New message with <em>Markdown</em>!</p>"`;

exports[`<Message /> should render a <strong> element 1`] = `"<p>New message with <strong>Markdown</strong>!</p>"`;
